File: telemetry-auth/src/main/java/rocks/telemetry/auth/configuration/PasswordHasher.java
package rocks.telemetry.auth.configuration;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.math.BigInteger;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;

@Component
public class PasswordHasher {

    private static final Logger log = LoggerFactory.getLogger(PasswordHasher.class);

    private static final int ITERATIONS_COUNT = 1000;
    private static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA1";
    private static final int HASH_BYTES = 24;

    public String generateHash(String password, String salt) {

        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), fromHex(salt), ITERATIONS_COUNT, HASH_BYTES * 8);

        try {
            SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
            return toHex(skf.generateSecret(spec).getEncoded());
        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
            log.warn("Error while pass hashing: ", e);
            return null;
        }
    }

    private String toHex(byte[] byteArray) {
        BigInteger bi = new BigInteger(1, byteArray);
        String hex = bi.toString(16);
        Integer paddingLength = byteArray.length * 2 - hex.length();
        if (paddingLength > 0)
            return String.format("%0" + paddingLength + "d", 0) + hex;
        else
            return hex;
    }

    private static byte[] fromHex(String hex) {
        byte[] binary = new byte[hex.length() / 2];
        for (int i = 0; i < binary.length; i++) {
            binary[i] = (byte) Integer.parseInt(hex.substring(2 * i, 2 * i + 2), 16);
        }
        return binary;
    }
}

File: telemetry-users/src/main/kotlin/rocks/telemetry/users/hashers/PasswordHasher.kt
package rocks.telemetry.users.hashers

import org.springframework.stereotype.Component
import java.math.BigInteger
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.security.spec.InvalidKeySpecException
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec


@Component
class PasswordHasher {

    companion object {

        val PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA1"
        val PBKDF2_ITERATIONS = 1000

        val SALT_BYTES = 24
        val HASH_BYTES = 24

        data class Password(val hash: String, val salt: String)
    }

    fun hash(password: String): Password {

        val random = SecureRandom()
        val salt = ByteArray(SALT_BYTES)
        random.nextBytes(salt)

        val hash = pbkdf2(password.toCharArray(), salt, PBKDF2_ITERATIONS, HASH_BYTES)

        return Password(hash.toHex(), salt.toHex())
    }

    @Throws(NoSuchAlgorithmException::class, InvalidKeySpecException::class)
    private fun pbkdf2(password: CharArray, salt: ByteArray, iterations: Int, bytes: Int): ByteArray {
        val spec = PBEKeySpec(password, salt, iterations, bytes * 8)
        val skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM)

        return skf.generateSecret(spec).encoded
    }
}

fun ByteArray.toHex(): String {
    val bi = BigInteger(1, this)
    val hex = bi.toString(16)

    val paddingLength = this.size * 2 - hex.length
    return if (paddingLength > 0)
        String.format("%0" + paddingLength + "d", 0) + hex
    else
        hex
}